# Features

## 1. long-term position control
## 2. B/S number per trade
## 3. KDJ risk control
## 4. trade log

In [1]:
import pandas as pd
import matplotlib.pyplot as plt
from main import Choice

In [2]:

choices = []

# 宽基指数
zz1000 = Choice(p_min=2.073, p_max=3.247, pos_max=20000,
                price=2.653, d_month=59, d_week=32, code="512100.SH")
kc50   = Choice(p_min=0.848, p_max=1.696, pos_max=50000,
                price=1.100, d_month=51, d_week=41, code="588000.SH")
gqhl   = Choice(p_min=1.740, p_max=2.135, pos_max=30000,
                price=1.866, d_month=55, d_week=47, code="501059.SH")
choices.append(zz1000)
choices.append(kc50)
choices.append(gqhl)

# 行业指数
yyetf  = Choice(p_min=0.454, p_max=0.666, pos_max=100000,
                price=0.443, d_month=32, d_week=18, code="512010.SH")
gfetf  = Choice(p_min=1.040, p_max=1.925, pos_max=30000,
                price=1.179, d_month=25, d_week=17, code="515790.SH")
hshl   = Choice(p_min=0.346, p_max=0.800, pos_max=100000,
                price=0.413, d_month=40, d_week=21, code="513330.SH")
nsdk   = Choice(p_min=0.776, p_max=1.500, pos_max=50000,
                price=1.043, d_month=62, d_week=93, code="513100.SH")
choices.append(yyetf)
choices.append(gfetf)
choices.append(hshl)
choices.append(nsdk)

# 转债银行
jyyh   = Choice(p_min=3.67, p_max=5.38, pos_max=13000,
                price=4.01, d_month=36, d_week=29, code="002807.SZ")
wxyh   = Choice(p_min=4.80, p_max=7.06, pos_max=10000,
                price=5.50, d_month=52, d_week=44, code="600908.SH")
hzyh   = Choice(p_min=8.88, p_max=16.43, pos_max=4000,
                price=11.94, d_month=25, d_week=52, code="600926.SH")
snyh   = Choice(p_min=4.41, p_max=6.77, pos_max=8000,
                price=4.59, d_month=29, d_week=32, code="603323.SH")
choices.append(jyyh)
choices.append(wxyh)
choices.append(hzyh)
choices.append(snyh)

512100.SH: buy_price_position
	price: 3.247, position: 0.000
	price: 2.777, position: 0.100
	price: 2.496, position: 0.300
	price: 2.256, position: 0.600
	price: 2.052, position: 1.000
512100.SH: sell_price_position
	price: 3.334, position: 0.000
	price: 2.928, position: 0.400
	price: 2.590, position: 0.700
	price: 2.308, position: 0.900
	price: 2.073, position: 1.000
588000.SH: buy_price_position
	price: 1.696, position: 0.000
	price: 1.357, position: 0.100
	price: 1.153, position: 0.300
	price: 0.980, position: 0.600
	price: 0.833, position: 1.000
588000.SH: sell_price_position
	price: 1.759, position: 0.000
	price: 1.465, position: 0.400
	price: 1.221, position: 0.700
	price: 1.018, position: 0.900
	price: 0.848, position: 1.000
501059.SH: buy_price_position
	price: 2.135, position: 0.000
	price: 1.977, position: 0.100
	price: 1.882, position: 0.300
	price: 1.802, position: 0.600
	price: 1.733, position: 1.000
501059.SH: sell_price_position
	price: 2.164, position: 0.000
	price: 2.0

## 1. long-term position control

In [3]:
for choice in choices:
    long_pos = choice.cal_position(choice.price, True)
    short_pos = choice.cal_position(choice.price, False)
    print(f"{choice.code}: price: {choice.price:.3f},"
          f"\n\tlong position: {long_pos:.3f}, {long_pos * choice.pos_max:.0f}"
          f"\n\tshort position: {short_pos:.3f}, {short_pos * choice.pos_max:.0f}")


512100.SH: price: 2.653,
	long position: 0.188, 3766
	short position: 0.644, 12874
588000.SH: price: 1.100,
	long position: 0.392, 19620
	short position: 0.819, 40951
501059.SH: price: 1.866,
	long position: 0.360, 10809
	short position: 0.801, 24025
512010.SH price 0.443 out of range: [0.450, 0.666]
512010.SH price 0.443 out of range: [0.454, 0.682]
512010.SH: price: 0.443,
	long position: 1.000, 100000
	short position: 1.000, 100000
515790.SH: price: 1.179,
	long position: 0.598, 17953
	short position: 0.921, 27644
513330.SH: price: 0.413,
	long position: 0.619, 61936
	short position: 0.926, 92621
513100.SH: price: 1.043,
	long position: 0.293, 14634
	short position: 0.759, 37967
002807.SZ: price: 4.010,
	long position: 0.537, 6981
	short position: 0.901, 11708
600908.SH: price: 5.500,
	long position: 0.374, 3739
	short position: 0.809, 8086
600926.SH: price: 11.940,
	long position: 0.262, 1049
	short position: 0.729, 2916
603323.SH: price: 4.590,
	long position: 0.783, 6266
	short p

## 2. buy/sell number per trade

#### 1. standard situation
set 1/40=2.5%

if trade while 3% decrease with 1/40 = 2.5% position, 
from top 100% price to bottom 50% price, we should trade about 23 times(0.97^23=0.51), we get 23*2.5% = 57.5% position

that seems ok because KDJ will hint us and the buy diff will decrease from 3% to 1%.

if trade every 1%, trade will happen about 68 times(0.99^68=0.50)

#### 2. scaled situation
set 1/40*scale=2.5%*scale, scale=ln(0.5)/ln(bottom_price)

from top 100% price to bottom 68% price, 3% diff will trade about 13 times($0.97^{13}=0.67$), we get $13*2.5% = 32.5%$ position; 1% diff will trade about 38 times($0.99^38=0.68$), we get $38*2.5%=95.0%$ position.

price range decreased $64%=(1-0.68)/(1-0.5)$
trade number decreased $43%=(23-13)/23$ and $44%=(68-38)/68$

To get similar position, we should trade more: 
trade number should increased: $77%=(23-13)/13$ and $79%=(68-38)/38$

summerize the theoretical formula:
- $0.97^{times_{standard}} = 0.5$
- $0.97^times_scaled   = bottom_price$
- $scale = times_standard / times_scaled$

so, we get:
- $ln(0.97) * times_standard = ln(0.5)$
- $ln(0.97) * times_scaled   = ln(bottom_price)$
- $scale = ln(0.5) / ln(bottom_price)$

for bottom_price = 0.68, $scale = ln(0.5) / ln(0.68) = 1.80$, consistent with the above calculation 77% and 79%.

In [4]:
for choice in choices:
    num_per_trade = choice.cal_pos_number_per_trade()
    print(f"{choice.code}: number_per_trade: {num_per_trade:.0f}")

512100.SH: number_per_trade: 772
588000.SH: number_per_trade: 1250
501059.SH: number_per_trade: 2541
512010.SH: number_per_trade: 4522
515790.SH: number_per_trade: 844
513330.SH: number_per_trade: 2067
513100.SH: number_per_trade: 1315
002807.SZ: number_per_trade: 589
600908.SH: number_per_trade: 449
600926.SH: number_per_trade: 113
603323.SH: number_per_trade: 323


## 3. KDJ risk control

In [5]:
for choice in choices:
    risk_density = choice.cal_risk_and_grid_density(choice.d_month, choice.d_week)
    print(f"{choice.code}: price: {choice.price:.3f},\n\t"
          f"risk level: {risk_density[0]}\n\t"
          f"buy sell contrast and diff: {risk_density[1:]}")

512100.SH: price: 2.653,
	risk level: 风险平衡
	buy sell contrast and diff: (0.03, 0.03, 0.03, 0.03)
588000.SH: price: 1.100,
	risk level: 风险平衡
	buy sell contrast and diff: (0.03, 0.03, 0.03, 0.03)
501059.SH: price: 1.866,
	risk level: 风险平衡
	buy sell contrast and diff: (0.03, 0.03, 0.03, 0.03)
512010.SH: price: 0.443,
	risk level: 很大机会
	buy sell contrast and diff: (0.03, 0.02, 0.04, 0.03)
515790.SH: price: 1.179,
	risk level: 巨大机会
	buy sell contrast and diff: (0.03, 0.015, 0.05, 0.03)
513330.SH: price: 0.413,
	risk level: 很大机会
	buy sell contrast and diff: (0.03, 0.025, 0.04, 0.03)
513100.SH: price: 1.043,
	risk level: 很大风险
	buy sell contrast and diff: (0.04, 0.03, 0.03, 0.02)
002807.SZ: price: 4.010,
	risk level: 风险平衡
	buy sell contrast and diff: (0.03, 0.03, 0.03, 0.03)
600908.SH: price: 5.500,
	risk level: 风险平衡
	buy sell contrast and diff: (0.03, 0.03, 0.03, 0.03)
600926.SH: price: 11.940,
	risk level: 很大机会
	buy sell contrast and diff: (0.03, 0.02, 0.04, 0.03)
603323.SH: price: 4.590,
	r

## 4. trade log

- add a trade transaction
- load from csv log
- save to csv log
